business/authentication/openshift_auth_controller.go

```diff
@@ -0,0 +1,191 @@
+	"errors"
+	"fmt"
+	"net/http"
+	"strconv"
+	"time"
+
+	"k8s.io/client-go/tools/clientcmd/api"
+
+	"github.com/kiali/kiali/business"
+	"github.com/kiali/kiali/config"
+	"github.com/kiali/kiali/log"
+)
+
+// openshiftSessionPayload holds the data that will be persisted in the SessionStore
+// in order to be able to maintain the session of the user across requests.
+type openshiftSessionPayload struct {
+	// Token is the access_token that was provided by the OpenShift OAuth server.
+	// It can be used against the cluster API.
+	Token string `json:"token,omitempty"`
+}
+
+// openshiftAuthController contains the backing logic to implement
+// Kiali's "openshift" authentication strategy. This authentication
+// strategy is basically an implementation of OAuth's implicit flow
+// with the specifics of OpenShift.
+type openshiftAuthController struct {
+	// businessInstantiator is a function that returns an already initialized
+	// business layer. Normally, it should be set to the business.Get function.
+	// For tests, it can be set to something else that returns a compatible API.
+	businessInstantiator func(authInfo *api.AuthInfo) (*business.Layer, error)
+
+	// SessionStore persists the session between HTTP requests.
+	SessionStore SessionPersistor
+}
+
+// NewOpenshiftAuthController initializes a new controller for handling OpenShift authentication, with the
+// given persistor and the given businessInstantiator. The businessInstantiator can be nil and
+// the initialized contoller will use the business.Get function.
+func NewOpenshiftAuthController(persistor SessionPersistor, businessInstantiator func(authInfo *api.AuthInfo) (*business.Layer, error)) *openshiftAuthController {
+	if businessInstantiator == nil {
+		businessInstantiator = business.Get
+	}
+
+	return &openshiftAuthController{
+		businessInstantiator: businessInstantiator,
+		SessionStore:         persistor,
+	}
+}
+
+// Authenticate handles an HTTP request that contains the access_token, expires_in URL parameters. The access_token
+// should be the token that was obtained from the OpenShift OAuth server and expires_in is the expiration date-time
+// of the token. The token is validated by obtaining the information user tied to it. Although RBAC is always assumed
+// when using OpenShift, privileges are not checked here.
+func (o openshiftAuthController) Authenticate(r *http.Request, w http.ResponseWriter) (*UserSessionData, error) {
+	err := r.ParseForm()
+
+	if err != nil {
+		return nil, fmt.Errorf("error parsing form info: %w", err)
+	}
+
+	token := r.Form.Get("access_token")
+	expiresIn := r.Form.Get("expires_in")
+	if token == "" || expiresIn == "" {
+		return nil, errors.New("token is empty or invalid")
+	}
+
+	expiresInNumber, err := strconv.Atoi(expiresIn)
+	if err != nil {
+		return nil, fmt.Errorf("token is empty or invalid: %w", err)
+	}
+
+	expiresOn := time.Now().Add(time.Second * time.Duration(expiresInNumber))
+	bs, err := o.businessInstantiator(&api.AuthInfo{Token: ""})
```

Comment:
Yes, empty is OK. I reckon this is confusing.

A following line is `user, err := bs.OpenshiftOAuth.GetUserInfo(token)`. That `GetUserInfo` func uses the received token to authenticate to the cluster API, rather than using whatever AuthInfo was passed in the business construction. I don't know why it was done in that way :-/